### DISCUSSION TITLE
Performance improvements for .NET layer

### DISCUSSION THREAD

**Discussion by bollhals** [google/or-tools#3064] [CP-SAT questions] [answered]
Hey

I'm using or-tools with .NET and while using it, I was looking into how exactly it was implemented and saw that there are a few things done in a suboptimal way. So I was thinking of contributing some pull requests that would improve things especially on the performance and allocation side.

As an example, there are various places in the code where it create a Domain class, just to then call the [IntVar ctor](https://github.com/google/or-tools/blob/stable/ortools/sat/csharp/IntegerExpressions.cs#L604) which calls `domain.FlattenedIntervals()`, we could add a new ctor with a `long lb, long ub` and directly use those instead of the overhead of creating the domain. (Speeds up the creation, avoids a lot of interop calls and results in exactly the same outcome)

Would you be interested in such Pull requests (targeted for .NET only)? I'm happy to work on them and explain the impact and gains from them.

**Comment by lperron** [comment]
Have you profiled the code ? We will only work on performance improvements if it is visible on a profile report. This being said, if it improves speed noticeably, we will gladly accept PR :-)

Thanks

**Reply by bollhals** [reply]
> Have you profiled the code ? We will only work on performance improvements if it is visible on a profile report. 

I've done some initial profiling and in my quick tests for the SAT nurses case I saw a +- 10% speedup and 30-40% .Net allocation reduction. 
Is there a prefered tool I should be using or a already defined way to profile it? 
(I've locally modified the examples to be able to run it as a .Net benchmark.)

**Comment by lperron** [comment, marked as answer]
But not on total time. Just on model building time.

Le sam. 8 janv. 2022, 12:28, Sandro Bollhalder ***@***.***> a
écrit :

**Reply by lperron** [reply]
Anyway, I pushed the improvement.

**Reply by bollhals** [reply]
It was the Total time without the reporting (as that one writes to console, which is bad for perf testing) but with all my local changes (there are a few other ideas 😀) I'll open a PR in the comming days and we can discuss there about the specifics. 

One thing to mention for the domain commit you did, in the c++ code you check that lb <= ub, and if not you don't add anything. That case should probably be also added to the change you did.

**Comment by lperron** [comment]
Actually not necessarily. The model will be invalid if you swap bounds.

Le sam. 8 janv. 2022, 17:28, Sandro Bollhalder ***@***.***> a
écrit :
